src/services/giant_api.rs

```diff
@@ -13,171 +14,222 @@ use crate::{
-        // Insert collection
-        let create_collection = CreateCollection {
-            name: collection.to_owned(),
-        };
-        let url = format!("{uri}/api/collections");
-        let res = client.post(url).json(&create_collection).send()?;
+    pub fn get_or_insert_collection(&mut self, ingestion_uri: &Uri) -> Result<Collection, CliError> {
+        let collection = ingestion_uri.collection();
+
+        let mut collections_url = self.base_url.clone();
+        collections_url.path_segments_mut()
+            .unwrap()
+            .push("api")
+            .push("collections");
+
+        let mut collection_url = collections_url.clone();
+        collection_url.path_segments_mut().unwrap().push(collection);
+
+        let res = self.send_request(self.client.get(collection_url))?;
         let status = res.status();
 
         if status == StatusCode::UNAUTHORIZED {
             return Err(CliError::APIAuthError);
-        } else if status != StatusCode::CREATED {
-            return Err(CliError::UnexpectedResponse(status));
         }
-        Ok(res.json::<Collection>()?)
-    } else {
-        Ok(res.json::<Collection>()?)
+
+        if status == StatusCode::NOT_FOUND {
+            // Insert collection
+            let create_collection = CreateCollection {
+                name: collection.to_owned(),
+            };
+            let res = self.send_request(self.client.post(collections_url).json(&create_collection))?;
+            let status = res.status();
+
+            if status == StatusCode::UNAUTHORIZED {
+                return Err(CliError::APIAuthError);
+            } else if status != StatusCode::CREATED {
+                return Err(CliError::UnexpectedResponse(status));
+            }
+            Ok(res.json::<Collection>()?)
+        } else {
+            Ok(res.json::<Collection>()?)
+        }
     }
-}
 
-pub fn get_or_insert_ingestion(
-    uri: &str,
-    ingestion_uri: &Uri,
-    base_collection: &Collection,
-    path: PathBuf,
-    languages: Vec<Language>,
-) -> Result<(), CliError> {
-    let collection = ingestion_uri.collection();
-    let ingestion = ingestion_uri.ingestion();
-
-    if base_collection
-        .ingestions
-        .iter()
-        .any(|i| i.uri == ingestion_uri.as_str())
-    {
-        // collection already contains ingestion!
-        Ok(())
-    } else {
-        let client = get_client(uri)?;
-        let url = format!("{uri}/api/collections/{collection}");
-
-        let create_ingestion = CreateIngestion {
-            path: Some(path),
-            name: Some(ingestion.to_owned()),
-            languages,
-            fixed: Some(false), // This is hardcoded to false in the existing CLI
-            default: Some(false),
+    pub fn get_or_insert_ingestion(
+        &mut self,
+        ingestion_uri: &Uri,
+        base_collection: &Collection,
+        path: PathBuf,
+        languages: Vec<Language>,
+    ) -> Result<(), CliError> {
```

Comment:
Agreed! This is Sam's original implementation - I noticed the same thing and had a comment against it which I removed in https://github.com/guardian/giant-utils/pull/4/commits/42a6ba715877f03e2c7526aeb1080e31d2ce6dc7

I will probably fix at some point. I think it should have type `Result<Ingestion, CliError>` because `get_or_insert_collection` has return type `Result<Collection, CliError>`